onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+
+Your function should return the final trained weight `m`.
+
+Additionally, in the "select one or more `(x, ylabel)` input pairs" step, for
+now you should just select the next `(x, ylabel)` pair in the dataset, and it
+should wrap back to the previous dataset.  So if your dataset is of length 10,
+then your eleventh iteration should reuse the first `(x, ylabel)` pair of your
+dataset.
+
+Here are some test cases:
+
+```python
+dataset = [(1,1)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=1)
+assert mfinal == XXX, mfinal
+
+dataset = [(1,1)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=100)
+assert mfinal == XXX, mfinal
+
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=4)
+assert mfinal == XXX, mfinal
+
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=100)
+assert mfinal == XXX, mfinal
+```
+
+### Q6
+
+What happens if you set the learning rate way too low?
+
+```python
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.000000001, dataset=dataset, iterations=4)
+print("mfinal", mfinal)
+```
+
+Explain in words what is happening.
+
+### Q7
+
+What happens if you set the learning rate way too high?
+
+```python
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=10, dataset=dataset, iterations=4)
+print("mfinal", mfinal)
+```
+
+Explain in words what is happening.
+
+## Batch Size in Training
+
+Your training fuction calculated the gradient for each data point, taking a
+step to minimize the loss for that particular point, then moving on to the next
+point.  This strategy (considering exactly one batch at a time) is called
+"stochastic gradient descent".  It can be an messy process because real gradient
+curves are not smooth parabolas like in this example, and taking a step every
+data point can cause you to take a lot of steps in the overall wrong direction.
+
+There is also "batch gradient descent", which is where you calculate the total
+gradient for your whole dataset (adding up the gradients for every data point
+in the whole dataset) before taking a single step.  This is a more reliable way
+to minimize the loss than stoachastic gradient descent because by looking at
+all data points before stepping, you can be sure you're stepping in the overall
+right direction.  However, nobody in deep learning really does this, because the
+datasets are normally huge and it would take forever to train a model on a
+dataset if you could only step once per epoch.
+
+There is an in-between option, which is more accurate than stochastic gradient
+descent but much fatser than batch gradient descent.  It is called "mini-batch
+gradient descent", and it's what basically the only gradient descent used in deep
+learning.  In mini-batch gradient descent, you have have some `batch_size`,
+usually a multiple of 2.  In each iteration of training, you select a
```

Comment:
Lol, yeah definitely not a multiple.

I'll downgrade it to "frequently a power of two" anyway, since I see that most of the time but not all of the time.